src/firmware/app/control/trajectory_planner.c

```diff
@@ -421,19 +562,106 @@ void app_trajectory_planner_generateVelocityTrajectory(
-    velocity_elements[last_element_index].angular_velocity  = 0;
+    velocity_elements[last_element_index].angular_velocity =
+        0.0;  // Final angular velocity is always assumed to be zero
     velocity_elements[last_element_index].time =
         position_elements[last_element_index].time;
 }
+
+TrajectoryPlannerGenerationStatus_t
+app_trajectory_planner_generateConstantInterpolationVelocityTrajectory(
+    FirmwareRobotPathParameters_t path_parameters,
+    VelocityTrajectory_t* velocity_trajectory, const float interpolation_period)
+{
+    PositionTrajectoryElement_t position_elements[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    PositionTrajectory_t position_trajectory;
+    position_trajectory.trajectory_elements = position_elements;
+
+    TrajectoryPlannerGenerationStatus_t status =
+        app_trajectory_planner_generateConstantInterpolationPeriodPositionTrajectory(
+            path_parameters, &position_trajectory, interpolation_period);
+    if (status != OK)
+    {
+        return status;
+    }
+
+    // Now calculate the velocity trajectory from the position trajectory
+    app_trajectory_planner_generateVelocityTrajectory(&position_trajectory,
+                                                      velocity_trajectory);
+    return status;
+}
+
+
+void app_trajectory_planner_generateStatesAndReturnSegmentLengths(
+    PositionTrajectory_t* trajectory,
+    TrajectorySegment_t trajectory_segments[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS])
+{
+    // Check that the pre conditions are met
+    assert(trajectory->path_parameters.num_segments > 2);
+    assert(trajectory->path_parameters.max_allowable_linear_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_linear_speed >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_speed >= 0);
+    assert(trajectory->path_parameters.initial_linear_speed >= 0);
+    assert(trajectory->path_parameters.final_linear_speed >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_speed >= 0);
+    assert(trajectory->path_parameters.num_segments <=
+           TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS);
+
+    const float t_start = trajectory->path_parameters.t_start;
+
+    PositionTrajectoryElement_t* trajectory_elements = trajectory->trajectory_elements;
+
+    trajectory_elements[0].orientation = shared_polynomial1d_getValueOrder3(
+        trajectory->path_parameters.orientation_profile, t_start);
+
+    trajectory_elements[0].position =
+        shared_polynomial2d_getValueOrder3(trajectory->path_parameters.path, t_start);
+
+    const float t_segment_size =
+        (trajectory->path_parameters.t_end - trajectory->path_parameters.t_start) /
+        (trajectory->path_parameters.num_segments - 1);
+
+    for (unsigned int i = 1; i < trajectory->path_parameters.num_segments; i++)
+    {
```

Comment:
We could clean up a few arguments here by adding:
```
const float current_t = t_start + i * t_segment_size;
```
and then using it instead of the entire expression below.

Reply:
Changed 👍